File: src/REXOS/MAS/HAL/ModuleIdentifier.java
package HAL;

import com.google.gson.JsonObject;
/**
 * This class provides a method of identifying modules
 * @author Tommas Bakker
 *
 */
public class ModuleIdentifier {
	private String manufacturer;
	private String typeNumber;
	private String serialNumber;
	
	public ModuleIdentifier(String manufacturer, String typeNumber, String serialNumber){
		this.manufacturer = manufacturer;
		this.typeNumber = typeNumber;
		this.serialNumber = serialNumber;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		ModuleIdentifier other = (ModuleIdentifier) obj;
		if (manufacturer == null) {
			if (other.manufacturer != null)
				return false;
		} else if (!manufacturer.equals(other.manufacturer))
			return false;
		if (serialNumber == null) {
			if (other.serialNumber != null)
				return false;
		} else if (!serialNumber.equals(other.serialNumber))
			return false;
		if (typeNumber == null) {
			if (other.typeNumber != null)
				return false;
		} else if (!typeNumber.equals(other.typeNumber))
			return false;
		return true;
	}

	public String getManufacturer(){
		return this.manufacturer;
	}
	public String getTypeNumber(){
		return this.typeNumber;
	}
	public String getSerialNumber(){
		return this.serialNumber;
	}
	
	public JsonObject toJSON(){
		JsonObject moduleIdentifier = new JsonObject();
		moduleIdentifier.addProperty("manufacturer", manufacturer);
		moduleIdentifier.addProperty("typeNumber", typeNumber);
		moduleIdentifier.addProperty("serialNumber", serialNumber);
		return moduleIdentifier;
	}
	
	/*
	@Deprecated
	public boolean equals(ModuleIdentifier rhs) {
		if(this.manufacturer.equals(rhs.getManufacturer()) == false) return false;
		if(this.typeNumber.equals(rhs.getTypeNumber()) == false) return false;
		if(this.serialNumber.equals(rhs.getSerialNumber()) == false) return false;
		return true;
	}
	*/
	
	@Override
	public String toString() {
		return manufacturer + " " + typeNumber + " " + serialNumber;
	}
	
}

File: src/REXOS/ROS/libraries/rexos_datatypes/src/ModuleIdentifier.cpp
#include <rexos_datatypes/ModuleIdentifier.h>

namespace rexos_datatypes{
	ModuleIdentifier::ModuleIdentifier() :
			ModuleTypeIdentifier("", ""), serialNumber("")
	{
		// nothing to do here
	}
	ModuleIdentifier::ModuleIdentifier(std::string manufacturer, std::string typeNumber, std::string serialNumber) :
			ModuleTypeIdentifier(manufacturer, typeNumber), serialNumber(serialNumber)
	{
		// nothing to do here
	}
	std::string ModuleIdentifier::getSerialNumber() const {
		return serialNumber;
	}
	std::string ModuleIdentifier::toString() {
		return "{\"manufacturer\":" + getManufacturer() + "\"typeNumber\":" + getTypeNumber() + "\"serialNumber\":" + getSerialNumber() + "}";
	}
	Json::Value ModuleIdentifier::toJSONObject() {
		Json::Value output;
		output["manufacturer"] = getManufacturer();
		output["typeNumber"] = getTypeNumber();
		output["serialNumber"] = getSerialNumber();
		return output;
	}
	bool ModuleIdentifier::operator==(const ModuleIdentifier& rhs) const {
		if(		this->getManufacturer() == rhs.getManufacturer() &&
				this->getTypeNumber() == rhs.getTypeNumber() &&
				this->getSerialNumber() == rhs.getSerialNumber()) {
			return true;
		} else {
			return false;
		}
	}
	std::ostream& operator<<(std::ostream& os, const ModuleIdentifier& obj) {
		os << "manufacturer: " << obj.getManufacturer();
		os << "typeNumber: " << obj.getTypeNumber();
		os << "serialNumber: " << obj.getSerialNumber();
		return os;
	}
}